lib/ansible/galaxy/collection.py

```diff
@@ -887,7 +888,8 @@ def _build_collection_tar(b_collection_path, b_tar_path, collection_manifest, fi
                 b_src_path = os.path.join(b_collection_path, to_bytes(filename, errors='surrogate_or_strict'))
 
                 def reset_stat(tarinfo):
-                    tarinfo.mode = 0o0755 if tarinfo.isdir() else 0o0644
+                    existing_is_exec = tarinfo.mode & stat.S_IEXEC
```

Comment:
This may be more clear using `S_IXUSR` instead of its synonym, `S_IEXEC`:

```suggestion
                    existing_is_exec = tarinfo.mode & stat.S_IXUSR
```

See https://docs.python.org/3/library/stat.html#stat.S_IEXEC